scopes/preview/preview/pre-bundle-preview.task.ts

```diff
@@ -0,0 +1,108 @@
+import { join } from 'path';
+import { existsSync, mkdirSync, writeFileSync } from 'fs';
+import { BuildContext, BuildTask, BuiltTaskResult, TaskLocation } from '@teambit/builder';
+import { Capsule } from '@teambit/isolator';
+import { Logger } from '@teambit/logger';
+import { UIRoot, UiMain } from '@teambit/ui';
+import { PreviewAspect } from '@teambit/preview';
+import { promisify } from 'util';
+import webpack from 'webpack';
+import { sha1 } from '@teambit/legacy/dist/utils';
+import createPreBundleConfig from './webpack/webpack.prebundle.config';
+import { getEntryForPreBundlePreview } from './pre-bundle-preview';
+
+export const PRE_BUNDLE_PREVIEW_TASK_NAME = 'PreBundlePreview';
+export const PRE_BUNDLE_PREVIEW_DIR = 'preview-pre-bundle';
+export const PRE_BUNDLE_PREVIEW_HASH_FILENAME = '.hash';
+
+async function build(uiMain: UiMain, logger: Logger, outputPath: string): Promise<webpack.Stats | undefined> {
+  logger.debug(`pre-bundle for preview: start`);
+  const ui = uiMain.getUi();
+  if (!ui) throw new Error('ui not found');
+  const [rootExtensionName, uiRoot] = ui;
+  const resolvedAspects = await uiRoot.resolveAspects('preview');
+  const mainEntry = getEntryForPreBundlePreview(resolvedAspects, rootExtensionName, 'preview', PreviewAspect.id);
+  const config = createPreBundleConfig(outputPath, mainEntry);
+
+  const compiler = webpack(config);
+  logger.debug(`pre-bundle for preview: running webpack`);
+  const compilerRun = promisify(compiler.run.bind(compiler));
+  const results = await compilerRun();
+
+  logger.debug(`pre-bundle for preview: completed webpack`);
+  if (!results) throw new Error('unknown error during pre-bundle for preview');
+  if (results?.hasErrors()) {
+    clearConsole();
+    throw new Error(results?.toString());
+  }
+
+  return results;
+}
+
+async function createBundleUiHash(uiRoot: UIRoot): Promise<string> {
+  const aspects = await uiRoot.resolveAspects('preview');
+  aspects.sort((a, b) => ((a.getId || a.aspectPath) > (b.getId || b.aspectPath) ? 1 : -1));
+  const aspectIds = aspects.map((aspect) => aspect.getId || aspect.aspectPath);
+  return sha1(aspectIds.join(''));
+}
```

Comment:
This is the same function as in the ui.main.runtime
just pass in the runtime name then you can re-use it

Reply:
I see. This is to avoid circular deps. Same to the function clearConsole below. I actually refactored it in the coming next PR to make the function standalone and reusable, with more refactors together. Is that OK?